src/uclua.1
.\"
.\" SPDX-License-Identifier: BSD-2-Clause
.\"
.\" Copyright (c) 2021 Kyle Evans <kevans@FreeBSD.org>
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd February 21, 2021
.Dt UCLUA 1
.Os
.Sh NAME
.Nm uclua
.Nd Lua to UCL bridge
.Sh SYNOPSIS
.Nm
.Op Fl -json | Fl -lua | Fl -ucl | Fl -yaml
.Op Fl o Ar output
.Op Fl s Ar sandbox
.Op Ar file ...
.Sh DESCRIPTION
The
.Nm
utility takes any given input files, processes them as lua configuration files,
then converts them to UCL.
.Pp
The given
.Ar file
list is specified in the order provided.
If
.Dq -
is specified as a
.Ar file ,
then stdin will be processed.
If no
.Ar file
is specified, then stdin will be processed.
Interactive processing of stdin is not supported.
.Pp
The following options are available:
.Bl -tag -width indent
.It Fl -json
Output the configuration as JSON.
.It Fl -lua
Output the configuration as Lua.
The primary benefit of this option is to reduce the input configuration to
fully resolve all variables.
Specifically, the output will have neither multiple definitions nor any
function definitions or function calls.
.It Fl -ucl
Output the configuration as UCL.
This is the default output format.
.It Fl -yaml
Output the configuration as YAML.
.It Fl o Ar output , Fl -output Ar output
Output the configuration to
.Ar output .
The default behavior is to write the configuration to stdout.
.It Fl s Ar sandbox , Fl -sandbox Ar sandbox
Sandbox directory to confine
.Fn require
to searching.
Directories within the
.Ar sandbox
will also be searched for scripts.
.El
.Sh SEE ALSO
.Xr libucl 3 ,
.Xr uclua 3
.Sh AUTHORS
The
.Nm
utility was written by
.An Kyle Evans Aq Mt kevans@FreeBSD.org .


src/uclua.c
/*-
 * SPDX-License-Identifier: BSD-2-Clause
 *
 * Copyright (c) 2021 Kyle Evans <kevans@FreeBSD.org>
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/param.h>

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

#include <getopt.h>

#include <uclua.h>

enum {
	JSON_OPT = CHAR_MAX + 1,
	LUA_OPT,
	UCL_OPT,
	YAML_OPT,
};

static const char *optstr = "o:s:";

static struct option longopts[] = {
	{ "json",	no_argument,	NULL,	JSON_OPT },
	{ "lua",	no_argument,	NULL,	LUA_OPT },
	{ "ucl",	no_argument,	NULL,	UCL_OPT },
	{ "yaml",	no_argument,	NULL,	YAML_OPT },
	{ "output",	required_argument,	NULL,	'o' },
	{ "sandbox",	required_argument,	NULL,	's' },
};

static int
usage(void)
{

	fprintf(stderr, "Usage: %s [--json | --lua | --ucl | --yaml] "
	    "[-o output] [-s sandbox] [file ...]\n", getprogname());
	return (1);
}

static int
parse_one(lcookie_t *lcook, const char *name)
{
	FILE *cfg;
	int ret;

	if (strcmp(name, "-") == 0)
		cfg = stdin;
	else
		cfg = fopen(name, "r");
	if (cfg == NULL) {
		printf("Failed to open file '%s'\n", name);
		return (1);
	}

	ret = 0;
	if (!uclua_parse_file(lcook, cfg)) {
		ret = 1;
		fprintf(stderr, "Failed to parse from %s\n", cfg == stdin ? "stdin" : name);
	}

	if (cfg != stdin)
		fclose(cfg);
	return (ret);
}

int
main(int argc, char *argv[])
{
	lcookie_t *lcook;
	FILE *outf;
	const char *outfile, *sandbox;
	char *cwd;
	int ch, ret;
	uclua_dump_type udump;

	udump = UCLUAD_UCL;
	cwd = NULL;
	sandbox = outfile = NULL;
	while ((ch = getopt_long(argc, argv, optstr, longopts, NULL)) != -1) {
		switch (ch) {
		case JSON_OPT:
			udump = UCLUAD_JSON;
			break;
		case LUA_OPT:
			udump = UCLUAD_LUA;
			break;
		case UCL_OPT:
			udump = UCLUAD_UCL;
			break;
		case YAML_OPT:
			udump = UCLUAD_YAML;
			break;
		case 'o':
			outfile = optarg;
			break;
		case 's':
			sandbox = optarg;
			break;
		default:
			usage();
		}
	}

	argc -= optind;
	argv += optind;

	if (argc == 0 && isatty(STDIN_FILENO)) {
		fprintf(stderr, "interactive conversion not supported\n");
		return (usage());
	}

	if (outfile == NULL || strcmp(outfile, "-") == 0) {
		outf = stdout;
	} else {
		outf = fopen(outfile, "w");
		if (outf == NULL) {
			fprintf(stderr, "could not open '%s' for output\n", outfile);
			return (usage());
		}
	}

	lcook = uclua_new();
	if (lcook == NULL) {
		fprintf(stderr, "out of memory\n");
		ret = 1;
		goto out;
	}

	if (sandbox == NULL)
		sandbox = cwd = getcwd(NULL, 0);
	if (sandbox != NULL) {
		if (!uclua_set_sandbox(lcook, sandbox)) {
			ret = 1;
			goto out;
		}

		free(cwd);
		cwd = NULL;
	}

	if (argc == 0) {
		ret = parse_one(lcook, "-");
	} else {
		ret = 0;
		for (int i = 0; i < argc; i++) {
			if ((ret = parse_one(lcook, argv[i])) != 0)
				break;
		}
	}

	if (ret == 0 && uclua_dump(lcook, udump, outf) != 0) {
		fprintf(stderr, "Failed to dump!\n");
		ret = 1;
	}
out:
	free(cwd);
	if (outf != stdout)
		fclose(outf);
	if (lcook != NULL)
		uclua_free(lcook);
	return (ret);
}
